### PR TITLE
User profiles

### PR DESCRIPTION
**Pull request by afrasier**
This PR implements the baseline user profile for the TalentMAP system.

Including @jseppi on the reviewer list as this is going to be the biggest chunk of code in the Sprint, so it might be good to review it now in this separate PR.

* Add `UserProfile` model which links to the default Django authorization `User` model, extending it to allow the storage of `favorite_positions`, `position_preferences`, and `language_qualifications`
* Added `create_user` management command to allow the creation of different users for development purposes
* Added basic authentication via Sessions and Expiring Tokens
* Added serializers to allow for the creation and updating of `UserProfile`, `Qualification` objects and a common mixin to support changing serializers based upon endpoint action
* Updated common serializer `PrefetchedSerializer` to pass overridden child fields on to children. For example, the `Position` object has a nested `Post` object. Prior to this change, one could not specify `post__description` in the field limitation, as it would not be passed to the child. This is now fixed.
* Add testing fixtures for `authorized_client` and `authorized_user` for use in testing authorization required endpoints
* Added tests for new endpoints

* Added the following endpoints (:lock: means the endpoint requires a valid session/token)
   * `/api/v1/accounts/` - Supports session logins
   * POST `/api/v1/accounts/token/` - Supports token logins (default expiration: 1 day)
   * GET `/api/v1/accounts/profile/` - :lock: - Displays the logged-in user's profile
   * PATCH `/api/v1/accounts/profile/` - :lock: - Updates the user profile with the data payload
   * PUT `/api/v1/language/qualifications/` -  :lock: - Creates a new language qualification
   * GET `/api/v1/position/preferred/` - :lock: - Lists all positions, filtered by user's stored preferences
   * GET `/api/v1/position/favorites/` - :lock: - Lists all of the current user's favorited positions

### CODE REVIEW

**Comment by afrasier** [comment]
Looks like @burgwyn is having some problems with the Expiring Tokens library; so I'll be moving that functionality into our codebase proper.

File: talentmap_api/common/serializers.py [hunk 1]
```diff
@@ -35,10 +35,30 @@ def __init__(self, *args, **kwargs):
...
 
                 # If our serializer field name is not the same as the source, specify it
-                if name != nested["field"]:
+                if nested.get("field", False) and name != nested["field"]:
                     kwargs["source"] = nested["field"]
                     self.fields.pop(nested["field"])
 
+                # If our parent serializer has field limitations on the child, pass them down
+                child_override_fields = []
+                child_override_exclude = []
```
Comment (jseppi) [inline talentmap_api/common/serializers.py]: any way you could test `PrefetchedSerializer` in isolation?

File: talentmap_api/user_profile/models.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    '''
+    Position preferences should be a JSON object of filters representing a user's position preferences.
+    For example, suppose our user preferred posts with post danger pay >= 20 and with grade = 05
+
+    {
+        "post__danger_pay__gte": 20,
+        "grade__code": "05"
+    }
+    '''
+    position_preferences = JSONField(default=dict, help_text="JSON object containing filters representing a user's position preferences")
```
Comment (jseppi) [inline talentmap_api/user_profile/models.py]: Should these preferences be validated in any way before being passed to a query? Or maybe before being saved? Looks like anything can be stuffed in there currently.

File: talentmap_api/user_profile/models.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+    favorite_positions = models.ManyToManyField('position.Position', related_name='favorited_by_users', help_text="Positions which this user has designated as a favorite")
+
+
+@receiver(post_save, sender=User)
```
Comment (jseppi) [inline talentmap_api/user_profile/models.py]: 👍

File: talentmap_api/common/serializers.py [hunk 4]
```diff
@@ -35,10 +35,30 @@ def __init__(self, *args, **kwargs):
...
+                    for field in overrides:
+                        split_field = field.split(LOOKUP_SEP)
+                        # Skip this field if we don't have a nested field
+                        if len(split_field) == 1:
+                            continue
+                        if split_field[0] == name or split_field[0] == nested.get("field", ""):
+                            child_overrides.append(LOOKUP_SEP.join(split_field[1:]))
+
+                    # If we have child overrides, attach them to the child's kwargs
+                    if len(child_overrides) > 0:
```
Comment (afrasier) [inline talentmap_api/common/serializers.py]: 👍

File: talentmap_api/common/serializers.py [hunk 5]
```diff
@@ -35,10 +35,30 @@ def __init__(self, *args, **kwargs):
...
+                if nested.get("field", False) and name != nested["field"]:
                     kwargs["source"] = nested["field"]
                     self.fields.pop(nested["field"])
 
+                # If our parent serializer has field limitations on the child, pass them down
+                child_override_fields = []
+                child_override_exclude = []
+
+                for pair in [(override_fields, "override_fields"), (override_exclude, "override_exclude")]:
+                    overrides = pair[0]
```
Comment (afrasier) [inline talentmap_api/common/serializers.py]: Yes, down when the list of overrides is passed into the kwargs

File: talentmap_api/user_profile/models.py (hunk 2 above)
Comment (afrasier) [inline talentmap_api/user_profile/models.py]: Good point, I will add that

File: talentmap_api/common/serializers.py (hunk 1 above)
Comment (afrasier) [inline talentmap_api/common/serializers.py]: Yeah, I'll split it out into some functions and test those in isolation

**Comment by afrasier** [comment]
@jseppi Fixed the issues with the comments, split out some of the common serializer functionality and tested that in isolation. Added some tests to the profile endpoints to the generic endpoint tester (Adding my own functionality for custom recipes) and added validation/testing to the position preferences.

File: talentmap_api/common/serializers.py (hunk 5 above)
Comment (jseppi) [inline talentmap_api/common/serializers.py]: oh, whoops, missed it 👍

File: talentmap_api/user_profile/serializers.py [hunk 6]
```diff
@@ -53,6 +56,20 @@ class Meta:
...
+        '''
+        Validate user profile here, mainly position preferences as this is an
+        arbitrary json object
+        '''
+        if 'position_preferences' in data:
+            # Verify that position preferences are a valid set of position filters
+            try:
+                Position.objects.filter(Q(**data.get('position_preferences')))
+            except:
+                raise ValidationError("Position preferences must be a valid set of position filters")
```
Comment (jseppi) [inline talentmap_api/user_profile/serializers.py]: 👍

File: talentmap_api/user_profile/tests/test_user_profile_endpoints.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+    assert resp.status_code == status.HTTP_200_OK
+    assert len(resp.data["favorite_positions"]) == 1
+    assert len(resp.data["language_qualifications"]) == 1
+    assert resp.data["position_preferences"] == {'post__danger_pay__gte': 20}
+
+    assert list(authorized_user.profile.favorite_positions.values_list("id", flat=True)) == [2]
+    assert list(authorized_user.profile.language_qualifications.values_list("id", flat=True)) == [2]
+
+    # Post a bad position preferences
```
Comment (jseppi) [inline talentmap_api/user_profile/tests/test_user_profile_endpoints.py]: 👍

**Pull request merged**